### PR TITLE
Add support for pypsexec and smbprotocol

### PR DESCRIPTION
**Pull request by dwoz**
### What does this PR do?

- Removes the need for winexe which is not packaged for all platforms we
  support. The replacement, pypsexec, is pure python so it will work
  everywhere. #46818 

- Add support for smbprotocol to be used instead of impacket.
  smbprotocol supports python 3 and is also used by pypsexec. #46819 

### What issues does this PR fix or reference?

#46818, #46819

### New Behavior

Installing pypsexec and smbprotocol allows us to spin up windows boxes using salt cloud under python 3 without the need for the winexe executable. This also makes using winrm under python 3 work.

### Tests written?

Yes

### Commits signed with GPG?

Yes

### CODE REVIEW

File: doc/man/salt.7 [hunk 1]
```diff
@@ -54744,27 +54744,19 @@ supported by Salt Cloud. However, it may not necessarily be available on all
...
-channels:
+More information on smbprotocol can be found at the project home:
 .INDENT 0.0
 .IP \(bu 2
-\fI\%RPMs at pbone.net\fP
-.UNINDENT
-.INDENT 0.0
-.IP \(bu 2
-\fI\%openSUSE Build Service\fP
+\fI\%smbprotocol project home\fP
```
Comment (gtmanfred) [inline doc/man/salt.7]: This file does not need to be edited, it is generated when the package is built.

**Review by gtmanfred — CHANGES_REQUESTED** [review]
Can you revert the changes to salt.7, and also add a whole section to the doc/topics/releases/fluorine.rst

Otherwise, this looks good to me

**Review by gtmanfred — APPROVED** [review]

**Comment by dmurphy18** [comment]
@dwoz will this still need impacket to be packaged by Salt Stack, or are the checks for in case it is already installed ?

**Comment by dwoz** [comment]
@dmurphy18 The checks are to warn users about the depreciation if they already have impacket installed. If smprotocol gets installed, it will be preferred over impacket.

**Comment by UtahDave** [comment]
I'm getting some errors when using this PR with salt-ssh.
First I tested with the very latest in develop on Python 2.7 and impacket.   salt-ssh worked correctly.

Then I checked out this PR and installed Salt and smbprotocol. I uninstalled impacket.

Here's the output I got:

```
[boucha@elsapo salt (dwoz-pypsexec)]$ sudo salt-ssh test-win-ssh disk.usage
[WARNING ] /usr/local/lib/python2.7/dist-packages/salt/utils/thin.py:549: DeprecationWarning: Use of 'salt.utils.fopen' detected. This function has been moved to 'salt.utils.files.fopen' as of Salt 2018.3.0. This warning will be removed in Salt Neon.

[WARNING ] Failed auth for mech NEGOEX (1.3.6.1.4.1.311.2.2.30): Mech Type 1.3.6.1.4.1.311.2.2.30 is not yet supported
[ERROR   ] An un-handled exception from the multiprocessing process 'MultiprocessingProcess-1' was caught:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/process.py", line 765, in _run
    return self._original_run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 514, in handle_routine
    stdout, stderr, retcode = single.run()
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 990, in run
    stdout, stderr, retcode = self.cmd_block()
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 1288, in cmd_block
    stdout, stderr, retcode = self.shim_cmd(cmd_str)
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 1246, in shim_cmd
    self.shell.send(shim_tmp_file.name, target_shim_file, makedirs=True)
  File "saltwinshell/core.pyx", line 127, in saltwinshell.core.Shell.send (saltwinshell/core.c:3567)
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/smb.py", line 359, in put_file
    password=password
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/smb.py", line 331, in _put_file_smbprotocol
    file_open = conn.open_file(tree, path)
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/smb.py", line 119, in open_file
    CreateOptions.FILE_NON_DIRECTORY_FILE,
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/open.py", line 1012, in create
    return self._create_response(request)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/open.py", line 1018, in _create_response
    response = self.connection.receive(request)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/connection.py", line 1037, in receive
    raise smbprotocol.exceptions.SMBResponseException(response, status)
SMBResponseException: Received unexpected status from the server: (3221225485) STATUS_INVALID_PARAMETER: 0xc000000d
Process MultiprocessingProcess-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/process.py", line 765, in _run
    return self._original_run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 514, in handle_routine
    stdout, stderr, retcode = single.run()
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 990, in run
    stdout, stderr, retcode = self.cmd_block()
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 1288, in cmd_block
    stdout, stderr, retcode = self.shim_cmd(cmd_str)
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 1246, in shim_cmd
    self.shell.send(shim_tmp_file.name, target_shim_file, makedirs=True)
  File "saltwinshell/core.pyx", line 127, in saltwinshell.core.Shell.send (saltwinshell/core.c:3567)
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/smb.py", line 359, in put_file
    password=password
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/smb.py", line 331, in _put_file_smbprotocol
    file_open = conn.open_file(tree, path)
  File "/usr/local/lib/python2.7/dist-packages/salt/utils/smb.py", line 119, in open_file
    CreateOptions.FILE_NON_DIRECTORY_FILE,
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/open.py", line 1012, in create
    return self._create_response(request)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/open.py", line 1018, in _create_response
    response = self.connection.receive(request)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/connection.py", line 1037, in receive
    raise smbprotocol.exceptions.SMBResponseException(response, status)
SMBResponseException: Received unexpected status from the server: (3221225485) STATUS_INVALID_PARAMETER: 0xc000000d
[ERROR   ] Target 'test-win-ssh' did not return any data, probably due to an error.
test-win-ssh:
    Target 'test-win-ssh' did not return any data, probably due to an error.
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
TypeError: 'str' object does not support item assignment
Traceback (most recent call last):
  File "/usr/local/bin/salt-ssh", line 10, in <module>
    salt_ssh()
  File "/usr/local/lib/python2.7/dist-packages/salt/scripts.py", line 425, in salt_ssh
    client.run()
  File "/usr/local/lib/python2.7/dist-packages/salt/cli/ssh.py", line 24, in run
    ssh.run()
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 791, in run
    data['jid'] = jid  # make the jid in the payload the same as the jid in the tag
TypeError: 'str' object does not support item assignment
Traceback (most recent call last):
  File "/usr/local/bin/salt-ssh", line 10, in <module>
    salt_ssh()
  File "/usr/local/lib/python2.7/dist-packages/salt/scripts.py", line 425, in salt_ssh
    client.run()
  File "/usr/local/lib/python2.7/dist-packages/salt/cli/ssh.py", line 24, in run
    ssh.run()
  File "/usr/local/lib/python2.7/dist-packages/salt/client/ssh/__init__.py", line 791, in run
    data['jid'] = jid  # make the jid in the payload the same as the jid in the tag
TypeError: 'str' object does not support item assignment
```

**Comment by rallytime** [comment]
@dwoz There are some lint failures here: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/20926/violations/

**Comment by dwoz** [comment]
@rallytime Thanks, I will look into them.

@UtahDave I will be testing the salt-ssh setup over the weekend.

**Comment by UtahDave** [comment]
@dwoz Let's test this after updating to the latest develop after this big salt-ssh PR:   https://github.com/saltstack/salt/pull/46684

**Comment by rallytime** [comment]
@dwoz Can you rebase this with latest develop and see how the tests go?

File: salt/utils/cloud.py [hunk 2]
```diff
@@ -779,15 +799,71 @@ def wait_for_port(host, port=22, timeout=900, gateway=None):
...
+
+    def __init__(self, server, username=None, password=None, port=445,
+                 encrypt=True, service_name=None):
+        super(Client, self).__init__(server, username, password, port, encrypt)
+        self.service_name = service_name
+        self._exe_file = "%s.exe" % self.service_name
+        self._service = ScmrService(self.service_name, self.session)
+
+    def remove_service(self, wait_timeout=10, sleep_wait=1):
+        """
```
Comment (rallytime) [inline salt/utils/cloud.py]: These need to be single quotes rather than double for salt coding standards :)

File: salt/utils/smb.py [hunk 3]
```diff
@@ -21,6 +27,114 @@
...
+        AccessAllowedAce, AccessMask, AclPacket, SDControl, SIDPacket,
+        SMB2CreateSDBuffer
+    )
+    logging.getLogger('smbprotocol').setLevel(logging.WARNING)
+    HAS_SMBPROTOCOL = True
+except:
+    HAS_SMBPROTOCOL = False
+
+
+class MissingSmb(SaltException):
```
Comment (rallytime) [inline salt/utils/smb.py]: Would it make sense to move this to the `exceptions.py` file instead?

**Review by rallytime — CHANGES_REQUESTED** [review]

**Comment by dwoz** [comment]
@UtahDave Can you test this again. I've am able to run salt-ssh against a brand new box with python 2 now.

**Comment by dwoz** [comment]
Tested this with python 3 but it's dependent on some changes to saltwinshell.

File: doc/topics/cloud/windows.rst [hunk 4]
```diff
@@ -9,6 +9,13 @@ Windows images.
 
 Requirements
 ============
+
+.. note::
+   Support ``winexe`` and ``impacket`` has been deprecated and will be removed in
+   Fluorine. These dependencies are replaced by `pypsexec` and `smbprotocol`
```
Comment (rallytime) [inline doc/topics/cloud/windows.rst]: The back-ticks around pypsexec and smbprotocol should be double back-ticks for these to format properly in the docs.

File: doc/topics/releases/fluorine.rst [hunk 5]
```diff
@@ -470,3 +470,19 @@ It is also possible to use several alternative versions of Salt. You can for ins
...
+Dependency Deprecations
+=======================
+
+Salt-Cloud has been updated to use the ``pypsexec`` python library instead of the
+``winexe`` executable. Both ``winexe`` and ``pypsexec`` run remote commands
+against Windows OSes. Since ``winexe`` is not packaged for every system, it has
+been deprecated in favor of ``pypsexec``.
+
+Salt-Cloud has deprecated the use ``impacket`` in favor of ``smbprotocol``.
+This changes was made because ``impacket`` is not compatible with python 3.
```
Comment (rallytime) [inline doc/topics/releases/fluorine.rst]: `python 3` --> `Python 3`

File: salt/utils/cloud.py [hunk 6]
```diff
@@ -779,15 +799,85 @@ def wait_for_port(host, port=22, timeout=900, gateway=None):
...
+      - Keep trying service and file deletes since they may not succeed on the
+        first try. Raises an exception if they do not succeed after a timeout
+        period.
+    '''
+
+    def __init__(self, server, username=None, password=None, port=445,
+                 encrypt=True, service_name=None):
+        self.service_name = service_name
+        self._exe_file = "{0}.exe".format(self.service_name)
+        self._client = PsExecClient(server, usernamem, password, prot, encrypt)
```
Comment (rallytime) [inline salt/utils/cloud.py]: `username` and `port` are misspelled here.

**Review by rallytime — CHANGES_REQUESTED** [review]

**Review by rallytime — APPROVED** [review]

**Comment by dwoz** [comment]
@UtahDave @rallytime I have no objects to this going to 2018.3.

**Comment by rallytime** [comment]
This is a pretty big change so I would usually prefer to put this in develop.

Is the current method for salt cloud not PY3 compatible? Or is there more information/context you can provide about why this would need to go into a bug-fix release?

I'd like @cachedout to weigh in here on this as well.

**Comment by gtmanfred** [comment]
the current method is not py3 compatible, because it requires impacket, which only works on python2.

And also winexe, which doesn't build anymore for us, so we had to remove it from the repositories.

**Review by UtahDave — APPROVED** [review]

**Pull request merged**